### DISCUSSION TITLE
Perform a Sub Query on Insert

### DISCUSSION THREAD

**Discussion by monkeber** [SeaQL/sea-orm#1482] [Q&A] [answered]
I'm a bit new to ORM libraries in general and previously used to have predefined SQL queries in code that I would construct by hand. I'm looking for how to do the following.

Suppose I have two tables, `balance_changes` and `currencies`, the latter being a lookup table, what I want to do is:
```sql
INSERT INTO balance_changes(currency_id, amount)
	VALUES
		((SELECT id FROM currencies WHERE name = 'EUR'), 5.0);
```
I.e. I want to make a sub query while inserting a value into some table. I've found examples of [how to do it for Select](https://www.sea-ql.org/SeaORM/docs/advanced-query/subquery/)  but can't figure out how to do it on Insert or if it's possible at all.

I've tried something like this (which obviously does not compile):
```rust
balance_changes::ActiveModel {
        amount: ActiveValue::Set(5.0),
        currency_id: currencies::Column::Id.in_subquery(
            Query::select()
                .expr(currencies::Column::Id)
                .from(currencies::Entity)
                .and_where(Expr::col(currencies::Column::Name).eq("EUR".to_owned()))
                .to_owned(),
        ),
        ..Default::default()
    }
    .save(&txn)
    .await?;
```

Another option would to retrieve the required `Id` beforehand via standalone Select query and then just use the value, but I'm curious if I can do it all in one go. 

Is something like this possible via ORM library or perhaps my approach is incorrect?

**Comment by tyt2y3** [comment, marked as answer]
I suggest you to use SeaQuery to build the insert statement (i.e. you can convert a `Select` into sub query with [`into_sub_query_statement`](https://docs.rs/sea-query/latest/sea_query/query/trait.QueryStatementBuilder.html#tymethod.into_sub_query_statement)) and use it as the value expression

**Reply by monkeber** [reply]
Finally got time to check it out and yes, it seems to be working for me, thank you!

My test code:
```rust
    let txn = get_transaction().await?;

    let balance_columns: Vec<Alias> = [
        balance_changes::Column::Amount.as_str(),
        balance_changes::Column::CurrencyId.as_str(),
    ]
    .into_iter()
    .map(Alias::new)
    .collect();

    let select_column = (
        currencies::Entity,
        Alias::new(currencies::Column::Id.as_str()),
    );
    let sub_select = SimpleExpr::SubQuery(
        None,
        Box::new(
            Query::select()
                .column(select_column)
                .from(currencies::Entity)
                .and_where(Expr::col(currencies::Column::Name).eq("EUR".to_owned()))
                .to_owned()
                .into_sub_query_statement(),
        ),
    );

    let mut insert_stmt = Query::insert();
    insert_stmt
        .into_table(balance_changes::Entity)
        .columns(balance_columns)
        .values_panic([(5.0).into(), sub_select]);

    let builder = txn.get_database_backend();
    txn.execute(builder.build(&insert_stmt)).await?;

    txn.commit().await?;
```

If you don't mind me asking, is inserting via `ActiveModel` a common thing or using `sea_query` is a preferred approach? (For tables and db structures in general that use foreign keys fairly often).

**Comment by tyt2y3** [comment]
Thank you for sharing your solution with us!

ActiveModel is meant to insert Values, so if you are inserting expressions, or more advanced, by sub query select, then SeaQuery is the recommended (and only) way.

**Reply by monkeber** [reply]
Thank you for the explanation!
